unit-testing/unit-testing.md

```diff
@@ -0,0 +1,169 @@
+# Unit Testing
+
+## What makes a good Unit-Test?
+
+1. F.I.R.S.T. Principles
+   * Fast
+   * Isolated
+   * Repeatable
+   * Self-Verifying
+   * Timely
+2. Tests "one thing"
+3. Readable / Maintainable
+
+**Fast**
+> Unittests should give near immediate feedback to changes to the production code.
+> Hundreds of tests should be executable within seconds.
+
+**Isolated**
+> Tests must not depend on other tests to be executed.
+> Each test must stand on its own.
+> The order of execution should be non-deterministic.
+> Tests must also be isolated from external influences!
+> File system or Operating System must not change the tests behaviour.
+
+**Repeatable**
+> Unless something changed in the production code a test must always get the same result when executed.
+
+**Self-Verifying**
+> Unittest either *pass* OR *fail*.
+> There must not be a manual decision involved!
+
+**Timely**
+> Tests should be created before production code (TDD).
+
+**Tests "one thing"**
+> There should be only on reason a unit test fails.
+> If a single test tests more than one thing, it has more than one reason to fail.
+
+**Readable / Maintainable**
```

Comment:
There is also a point to be made that a well written test documents/explains the code it tests.